EIPS/eip-7702.md

```diff
@@ -0,0 +1,97 @@
+category: Core
+created: 2024-05-07
+---
+
+## Abstract
+
+Add a new transaction type that adds a `contract_code` field and a signature, and converts the signing account (not necessarily the same as the `tx.origin`) into a smart contract wallet for the duration of that transaction. Intended to offer similar functionality to [EIP-3074](./eip-3074.md).
+
+## Motivation
+
+There is a lot of interest in adding short-term functionality improvements to EOAs, increasing the usability of applications and in some cases allowing improved security. Three particular applications include:
+
+* **Batching**: allowing multiple operations from the same user in one atomic transaction. One common example is an [ERC-20](./eip-20.md) approval followed by spending that approval, a common workflow in DEXes that requires two transactions today. Advanced use cases of batching occasionally involve dependencies: the output of the first operation is part of the input to the second operation.
+* **Sponsorship**: account X pays for a transaction on behalf of account Y. Account X could be paid in some other ERC-20 for this service, or it could be an application operator including the transactions of its users for free.
+* **Privilege de-escalation**: users can sign sub-keys, and give them specific permissions that are much weaker than global access to the account. For example, you could imagine a permission to spend ERC-20 tokens but not ETH, or to spend up to 1% of total balance per day, or to interact only with a specific application.
+
+[EIP-3074](./eip-3074.md) solves all of these use cases. However, it has forward-compatibility concerns:
+
+* It introduces two opcodes, `AUTH` and `AUTHCALL`, that would have no use in an "endgame account abstraction" world where eventually all users are using smart contract wallets (which seems like it must happen eventually, at the least because eventually quantum computers will break the ECDSA that EOAs use)
+* It leads to the development of an "invoker contract" ecosystem that would be separate from the "smart contract wallet" ecosystem, leading to possible fragmentation of effort.
+
+The purpose of this EIP is to enable all of the use cases of EIP-3074, without these two weaknesses.
+
+## Specification
+
+The keywords “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
+
+### Parameters
+
+* `FORK_BLKNUM` = `TBD`
+* `TX_TYPE` = `TBD`
+* `MAGIC` = `TBD`
+* `PER_CONTRACT_CODE_BASE_COST` = `5000`
+
+As of `FORK_BLOCK_NUMBER`, a new [EIP-2718](./eip-2718.md) transaction is introduced with `TransactionType` = `TX_TYPE(TBD)`.
+
+The [EIP-2718](./eip-2718.md) `TransactionPayload` for this transaction is
+
+```
+rlp([chain_id, nonce, max_priority_fee_per_gas, max_fee_per_gas, gas_limit, destination, data, access_list, [[contract_code, y_parity, r, s], ...], signature_y_parity, signature_r, signature_s])
```

Comment:
A small suggestion: when `y_parity` is 0 or 1, it's a temporary setting code. When it's 2 or 3, it's `secp256r1` and a permanent setting code. When it's 4 or 5, it's `secp256k1` and a permanent setting code (currently not supported).

Your idea is very similar to mine, but you chose the `tx type`, so there's no need to add an opcode.
https://github.com/ethereum/EIPs/pull/8493/files

Reply:
Add a new field, don't overload yParity.  We don't need more obscure parsing rules.

Reply:
> Add a new field, don't overload yParity. We don't need more obscure parsing rules.

OK, sir. Do you agree with my idea?

Reply:
1. If EIP7702 has already been implemented, the cost of adding a new algorithm isn't high.

2. The risk of permanently upgrading an EOA to a CA is significant (as EOAs carry historical baggage), but addresses generated by algorithms like secp256r1 are not EOAs, so permanently setting code carries no additional risk.

3. People need a more stable way to ensure deterministic deployments.

4. Better backward compatibility: When temporary code setting is prohibited and only permanent code setting is allowed, simply reject that type and enable a new type.

5. Prevent user errors: In the past, Ethereum used block heights to determine certain code behaviors, which wasn't user-friendly. We should directly reject user requests via a flag since upgrading an EOA to a CA is risky.

Reply:
> Add a new field, don't overload yParity. We don't need more obscure parsing rules.

Yeah, like this:
0x00: Temporary upgrade And secp256k1  
0x10: Permanent upgrade And secp256k1  
0x01: Temporary upgrade And secp256r1  
0x11: Permanent upgrade And secp256r1